syntax = "proto3";

import "google/protobuf/empty.proto";

package airmobisim;

service AirMobiSim{
    
    rpc Start (google.protobuf.Empty) returns (google.protobuf.Empty);

    rpc ExecuteOneTimeStep (google.protobuf.Empty) returns (ResponseQuery);

    rpc Finish (google.protobuf.Empty) returns (google.protobuf.Empty);

    rpc GetManagedHosts(google.protobuf.Empty) returns (UavList);

    rpc InsertWaypoints (WaypointList) returns (google.protobuf.Empty);

    rpc InsertWaypoint (Waypoint) returns (google.protobuf.Empty);

    rpc InsertUAV (StartUav) returns (google.protobuf.Empty);
    
    rpc getMaxUavId (google.protobuf.Empty) returns (Number);

    rpc DeleteUAV (Number) returns (google.protobuf.Empty);

    rpc getNumberCurrentUAV (google.protobuf.Empty) returns (Number);
    
    rpc SetDesiredSpeed (UavSetSpeed) returns (google.protobuf.Empty);

    rpc UpdateWaypoints (WaypointList) returns (google.protobuf.Empty);

    rpc DeleteWaypoint (WaypointList) returns (google.protobuf.Empty);       

    rpc GetMaxSimulationTime(google.protobuf.Empty) returns (Number);

    rpc getMaxSimulationSteps(google.protobuf.Empty) returns (DoubleNumber);
}

message WaypointList {
    int32 id = 1;
    repeated Waypoint waypoint = 2;
}

message Waypoint {
    sint32 uid = 1;
    int32 index = 2;
    double x = 3;
    double y = 4;
    double z = 5;
}

message UavList {
    repeated Uav uavs = 1;
}

message Uav {
    int32 id = 1;
    double x = 2;
    double y = 3;
    double z = 4;
    double speed = 5;
    double angle = 6;
}

message Response{
    int32 id = 1;
    double x = 2;
    double y = 3;
    double z = 4;
    double speed = 5;
    double angle = 6;
}

message ResponseQuery{
    repeated Response responses = 1;
}

message StartUav{
    int32 id = 1;
    repeated Coordinates coordinates = 2;
    double speed = 3;
    double angle = 4;
    int32 mobilityModel = 5;
    bool removeNode = 6;
}
}

message Coordinates{
    double x = 1;
    double y = 2;
    double z = 3;
}

message Number{
    int32 num = 1;
}
message DoubleNumber{
    double num = 1;
}

message UavSetSpeed{
    int32 id = 1;
    double speed = 2;
}
